src/compiler/core.ts

```diff
@@ -323,8 +323,12 @@ namespace ts {
         return keys;
     }
 
-    export function getProperty<T>(map: Map<T>, key: string): T {
-        return hasOwnProperty.call(map, key) ? map[key] : undefined;
+    export function getProperty<T>(map: Map<T>, key: string): T | undefined {
+        return hasProperty(map, key) ? map[key] : undefined;
+    }
+
+    export function getOrUpdateProperty<T>(map: Map<T>, key: string, makeValue: () => T): T {

```

Comment:
There are a lot of places that use code like this -- just search for `] ||` -- but they seem to be using numeric IDs (so are safe from default properties) or are safe for other reasons.
I'm also currently looking into using Maps as maps.